derive/src/utils.rs

```diff
@@ -112,13 +112,21 @@ pub fn variant_index(v: &Variant, i: usize) -> TokenStream {
 		None
 	});
 
-	// then fallback to discriminant or just index
-	index.map(|i| quote! { #i }).unwrap_or_else(|| {
-		v.discriminant
-			.as_ref()
-			.map(|(_, expr)| quote! { #expr })
-			.unwrap_or_else(|| quote! { #i })
-	})
+	// if no attribute, use the discriminant if there is one
+	if index.is_none() {
+		if let Some((_, Expr::Lit(ExprLit { lit: Lit::Int(disc_lit), .. }))) = &v.discriminant {
+			index = disc_lit.base10_parse().ok()
+		}
+	}
+
+	// fall back to the provided index if none is found.
+	let index = index.unwrap_or(i);
+
+	// output the index with no suffix (ie 1 rather than 1usize) so that these can be quoted into
+	// an array of usizes and will work regardless of what type the discriminant values actually
+	// are.
+	let s = proc_macro2::Literal::usize_unsuffixed(index);
```

Comment:
what about 
```rust
#[repr(i32)]
enum Test {
    A = -1i32,
    B = 1i32,
}
```

Reply:
Good point! Currently indexes that don't parse into usizes are silently not accepted and default to being 0, so the above compiles but eg:

```rust
#[repr(i32)]
enum Test {
    A = -1i32,
    B = 0i32,
}
```

Fails with an error about dupe indexes.

Setting `#[index = -1]` would also fail but with the error `"Index must be in 0..255"` (see `check_variant_attribute`).

Since SCALE encoding currently always uses u8 to encode indexes, meaning that they must be within this range, I'd be inclined to not worry about supporting things outside of 0..=255 (and instead complain like the above error if we see them). I'll tweak the PR to improve the error in this case!